### DISCUSSION TITLE
Make sensor expiry alerts customizable to the minute

### DISCUSSION THREAD

**Discussion by DmitriyShepelev** [NightscoutFoundation/xDrip#3666] [General]
When enabled, sensor expiry alerts go off 24, 12, 6, and 2 hours before sensor expiration. These thresholds are hard coded, offering users no customizability.

Making these alerts customizable to the minute enables users to minimize data lapses in between sensor sessions. Ideally, an alert would go off five to ten minutes before expiry, allowing users just enough time to prepare a replacement sensor.

Perusing the code, I see that polling to trigger this alert is only done when the screen is on and upon a charger connection change; there seems to be a method to poll hourly, but it's unused:
https://github.com/NightscoutFoundation/xDrip/blob/26ae3c7b112d810c04570104df726d4f31372c7e/app/src/main/java/com/eveningoutpost/dexdrip/alert/Poller.java#L40C1-L42C6

Supporting this feature would thus require polling every minute.

My proposal is to move the "Sensor expiry" alert into its own section under "Extra Alerts (xDrip+)" (like "PERSISTENT HIGH ALERT" and "FORECASTED LOW ALERT"), where users can input an integer specifying the number of minutes before sensor expiry at which the alarm will sound, as well as a customizable alarm sound.

**Comment by Navid200** [comment]
Thanks for the thoughts and suggestions.

The existing sensor expiry alert alerts you multiple times.  You can think of it as a reminder rather than an alert.  
The reason it is set to be silent is just that.  You get the reminder.  It is up to you to act or not.  Hence, it will be quiet because you may be asleep.

It only goes off when you unlock the screen or when you connect or disconnect the charger.  As a result, the time it goes off is not exact because you may neither unlock the screen nor change the charger exactly 6 hours before the expiry.  

When a sensor stops, you get an alert whether you want it or not.  You get a ping and vibration telling you that your sensor has stopped.

Considering all of that and the intent, would it be adequate if we add one more time to the reminder at 15 minutes to?

**Reply by DmitriyShepelev** [reply]
> The existing sensor expiry alert alerts you multiple times. You can think of it as a reminder rather than an alert.

Personally (and I imagine this affects many people), I'm overloaded with notifications, so minimizing this set to one is desirable and prevents notification fatigue. I don't find it useful to be reminded multiple times of an event less than or equal to 24 hours in the future.

> Hence, it will be quiet because you may be asleep.

There's an ever-increasing number of diabetics using closed-loop systems configured with xDrip, where the slightest data lapse may result in loss of blood sugar control. For this user base, a non-silent alert may be desirable. In my opinion, this alert should have silent, vibrate only, and sound options.

> It only goes off when you unlock the screen or when you connect or disconnect the charger. As a result, the time it goes off is not exact because you may neither unlock the screen nor change the charger exactly 6 hours before the expiry.

If I'm already interacting with my phone, this logic kind of defeats the point of an alarm (or reminder); I can just make a few keystrokes and check the sensor expiry myself. Polling should be done regardless of user-interaction to minimize users' cognitive burden with managing this disease.

**Reply by Navid200** [reply]
My mistake!  I believe it will not go off if the screen is locked unless you plug or unplug the charger.
The intent is that it will not trigger if you are sleeping.  

Let's not think of this as the only option for being notified.  As I said, xDrip notifies you when the sensor expires.  

There is always room for improvement.  Sometimes, it should not necessarily be "let's change this".  The other possibility is to add another option.
We have low glucose alerts.  But, we also have forecast low alert.  They complement each other.

We can talk about what is missing.

But, I would like to mention what could go wrong if we change this.
I use G7.  A G7 sensor runs for 10.5 days.
I always like to start a sensor in the morning.  When the 12 hour before expiry notification goes off, it is time for me to insert a new sensor.
But, there have been times I have not been able to start the sensor in the morning.  For example, I was late in returning from a trip and by the time I inserted my sensor, it was 2pm.
In 10 days, if there was only one notification, it would go off at the wrong time.  I like the fact that xDrip notifies me at different periods.  In that case, I am reminded and plan to start the sensor when it suits me and not forced by when the alert goes off.

It is my plan to open a PR to create subpages for each one of the alerts on the extra alerts page.  I cannot be sure that it will be approved.  But, it is my plan to propose it.  In that case, sensor expiry will get its own subpage as you suggest.

There has been a request to make the 12-hour silent alert, for G7, a non-silent alert to be the same as the notification you get from Dexcom app after 10 days running a sensor.
It has been my plan to open a PR to make it optional to make the 12 hour notification non-silent.
Again, I don't know if it will be approved.

The more complexity we add to xDrip, the more mess we will end up with every time there is a new release of Android.  The more possibility of different things breaking exists then.  So, I always suggest that we only make changes to xDrip when absolutely necessary.

Thanks

**Reply by DmitriyShepelev** [reply]
> xDrip notifies you when the sensor expires.

Right, but if you want to change a sensor at the same time every day, then this notification is too late because one cannot immediately change a sensor upon expiration. It takes maybe one to two minutes to change a sensor, which means that the sensor expiry alarm will go off at least two minutes later than before. The end result is that sensor changes will need to be made later and later with time, making it difficult to maintain a consistent sensor replacement schedule.

> The more complexity we add to xDrip, the more mess we will end up with every time there is a new release of Android. The more possibility of different things breaking exists then. So, I always suggest that we only make changes to xDrip when absolutely necessary.

I mean, the only new functionality would be the new polling logic; the remaining logic already exists elsewhere in the app, so I'm not sure how much new complexity is actually going to be added.

---

I'm happy to draft a PR either way. If it's rejected, that's fine, I'll just build a custom version of the app for myself. I just thought it might be useful for others.

**Reply by Navid200** [reply]
Oh, I never meant you shouldn't open a PR.  

xDrip also has reminders.  I'm not sure if you are aware.

My only point was that there is a fundamental difference between a glucose alert and a sensor expiry alert.
As a diabetic, I need to know when I am low so that I can treat myself.  It makes no difference what time it is.  All I need to know is that my glucose is low.
But, a sensor alert will only tell me when my current sensor expires.  I need more facts to decide when to start the next sensor.  For example, if I started my current sensor too late, I will need to start my next sensor before the current one expires.
In that case, I wouldn't want to wait for the sensor alert to go off before I start my next sensor.

Please feel free to open a PR.

**Reply by DmitriyShepelev** [reply]
FYI #3675

**Comment by Navid200** [comment]
As a single step to address your needs, would it help if you could choose which one of the several existing sensor expiry notes would be active and which ones not?

Currently, if you enable the alert, it would notify you 24 hours before, 12 hours before, 6 hours before and 2 hours before expiry.
Would it be useful if there were checkboxes where you could choose only one of those to trigger and the rest to remain silent?

I know you also would like the alert to trigger no matter what even if the phone is not unlocked or the charger is not plugged or unplugged.  Hopefully, I can address that too later.

But, would checkboxes help?

**Comment by Navid200** [comment]
As of the 2026.07.15 Nightly release, you can control the sound for the sensor expiry notification by editing the automatically created notification channel for it.  While editing the notification channel, you can enable or disable vibration and you can choose a different sound file for it.
